packages/gatsby-link/src/index.tsx

```diff
@@ -182,39 +261,55 @@ class GatsbyLink extends React.Component {
   }
 }
 
-GatsbyLink.propTypes = {
-  ...NavLinkPropTypes,
-  onClick: PropTypes.func,
-  to: PropTypes.string.isRequired,
-  replace: PropTypes.bool,
-  state: PropTypes.object,
-}
-
-const showDeprecationWarning = (functionName, altFunctionName, version) =>
+const showDeprecationWarning = (
+  functionName: string,
+  altFunctionName: string,
+  version: number
+): void =>
   console.warn(
     `The "${functionName}" method is now deprecated and will be removed in Gatsby v${version}. Please use "${altFunctionName}" instead.`
   )
 
-export default React.forwardRef((props, ref) => (
-  <GatsbyLink innerRef={ref} {...props} />
+/**
+ * This component is intended _only_ for links to pages handled by Gatsby. For links to pages on other
+ * domains or pages on the same domain not handled by the current Gatsby site, use the normal `<a>` element.
+ */
+const GatsbyLink = React.forwardRef<
+  HTMLAnchorElement,
+  GatsbyLinkProps<unknown>
+>((props: GatsbyLinkProps<unknown>, ref: React.Ref<HTMLAnchorElement>) => (
+  <GatsbyLinkInternal innerRef={ref} {...props} />
 ))
+type GatsbyLink<TState = unknown> = React.Component<
+  GatsbyLinkProps<TState>,
+  IGatsbyLinkState
+>
```

Comment:
does this not cause problems declaring the type as the same as the const above?

Reply:
See note above on TypeScript declaration merging. It doesn't quite cause answers, but it's also not the prettiest. But we're in a bit of a bind because of the previous index.d.ts declaring it this way and consts not allowing generic types.